io: add pager for append-only file abstraction

dotnwat opened the pull request:
Add io::pager which provides an append-only abstraction on top of the files using io::scheduler for the write-back process.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented:
Force push to add missing header https://github.com/redpanda-data/redpanda/compare/01618eccda20f3ab86c46de5c627a13333940a28..3a4a0092702d1032887bfa3c2ef53ce2f0982fc6

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/46152#018e3af6-0292-4b5a-963c-df7cf4651e70
<!--2b00d6e9-64d3-48df-aadb-b54eb844e093-->

andrwng commented on src/v/io/tests/pager_test.cc:
```diff
@@ -0,0 +1,335 @@
+  ::testing::Combine(
+    ::testing::Bool(), ::testing::ValuesIn(test_case_sizes())));
+
+TEST_P(AppendPagerTest, Append) {
```
It would be great to have a test that tested appends and concurrent reads, especially with a cache that evicts aggressively

dotnwat replied:
oh yeh good call. we'll need to add locking for the append/read case you mention, and also for eviction. none of that is here though in this version that assumes no concurrency.

dotnwat commented on src/v/io/pager.cc:
```diff
@@ -0,0 +1,231 @@
+         * the scheduler does the right thing if pages are marked more than once
+         * as dirty and resubmitted for write-back.
+         */
+        page->set_flag(page::flags::dirty);
```
the flags we want in the pager are faulting/dirty only. the reason read/write flags exists right now is that we don't have a separate data structure (yet, probably a few PRs down the line) that represents a request in the IO queue--the io-queue just has a queue of pages so we needed a flag to represent if the page was being read or written. those read/write flags will disappear when we have a request structure (e.g. read,write, fsync, falloc, etc...)

dotnwat commented on src/v/io/pager.h:
```diff
@@ -0,0 +1,111 @@
+     * should advance the starting offset to retrieve additional pages. Reads
+     * past the end-of-file will return an empty result.
+     *
+     * Caller should ensure that the length of the read does not result in a
+     * vector that would exceed recommended contiguous allocation limits.
```
the comment was intended to prevent this kind of question haha. but no, you're right we might want to ditch the vector. on the other hand, we want callers to avoid holding up lots and lots of pages from being evictable by holding onto refereces. so in reality we'll probably want to impose a max amount of pages that we return from read(), in which case, a vector will probably be fine.

dotnwat commented on src/v/io/page.h:
```diff
@@ -57,8 +57,10 @@ class page {
 
     /*
      * read,write: page is queued for read or write
+     * faulting: page is being faulted in from storage
```
fixed

dotnwat commented on src/v/io/page.h:
```diff
@@ -88,6 +89,23 @@ class page {
+     * locations where a page must be tested to those where new page references
+     * are obtained from an index.
+     */
+    [[nodiscard]] bool may_evict() const {
```
fundamentally, no. a reader could cause start a page fault and then leave (e.g. cancel the read / timeout) while the background hydration is occurring so that the cache was the only reference to the page. in reality the io-queue holds a reference too, physically in the current implementation the implication you describe is present.

dotnwat commented on src/v/io/page_cache.h:
```diff
@@ -0,0 +1,57 @@
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "io/page.h"
```
fixed

dotnwat commented on src/v/io/pager.h:
```diff
@@ -0,0 +1,111 @@
+    size_t size_;
+    page_cache* cache_;
+    scheduler* sched_;
+    scheduler::queue queue_;
+    page_set pages_;
```
yep, fixed. always forget about class member variables.

dotnwat commented on src/v/io/pager.cc:
```diff
@@ -0,0 +1,231 @@
+        page.signal_waiters();
+    } else if (page.test_flag(page::flags::dirty)) {
+        vassert(page.test_flag(page::flags::write), "Expected write page");
+        if (!page.test_flag(page::flags::queued)) {
```
the io-queue clears the queued flag before write-back happens. normally the write finishes and the completion handler is called and the flag is still cleared.

but while the write was happening, the page may have been re-queued because new data arrived and was written into the page, in which case the queued flag will be set. this is also the case here in which we don't want to incorrectly clear the dirty flag.

now that i'm thinking about it, it might be simpler to only invoke the completion handler for writes when the page wasn't re-queued.

dotnwat replied:
fixed

andrwng replied:
Thanks for explaining, that's really helpful. I agree, it may make for easier reasoning to have the io-queue only call the completion callback when the page isn't requeued. Does this result in a meaningful behavioral change for users? Is it that writers now have to wait longer in the face of concurrent writers, before getting a page as we're appending? (I think that makes sense, if so)

dotnwat replied:
> Does this result in a meaningful behavioral change for users? Is it that writers now have to wait longer in the face of concurrent writers, before getting a page as we're appending? (I think that makes sense, if so)

For writers, there should be no change: new writes can always land into a dirty page as appends. That's true even if the page is queued for write-back.

For readers, there should be no change: dirty data in the cache is always readable.

andrwng replied:
Ah makes sense. I saw us calling `get_page()` on the write path and assumed that that could lead us to waiting on the write path. But it makes sense that we could return a dirty page, in which case we aren't faulting (and not waiting!)

dotnwat replied:
yep. i think that there is a good argument for having a fast path for `get_page` that doesn't use `co_await` which would make it more clear.

dotnwat commented on src/v/io/pager.cc:
```diff
@@ -0,0 +1,231 @@
+        vassert(page.test_flag(page::flags::read), "Expected read page");
+        page.clear_flag(page::flags::faulting);
+        page.signal_waiters();
+    } else if (page.test_flag(page::flags::dirty)) {
```
>Are faulting and dirty meant to be mutually exclusive?

yeh, these flags should mutually exclusive. i've added a check to catch violation so this invariant!

dotnwat commented:
Force-push addressed feedback https://github.com/redpanda-data/redpanda/compare/3a4a0092702d1032887bfa3c2ef53ce2f0982fc6..f224a3985bd0d0a1e2a156051f616147ea43ef56

dotnwat commented:
Force pushed https://github.com/redpanda-data/redpanda/compare/f308e8a02052f3625a4bfd81ba9871515128b3d0..ae5efe90381fb29d1c0ad284f174ac805db8c228 to fix Git commit messages and add a note about the pager into the io/README

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/46902#018e8115-18bc-4a5e-a666-84e734b6968e:
```
"rptest.tests.transactions_test.TxUpgradeTest.upgrade_does_not_change_tx_coordinator_assignment_test"
```
<!--09bffe7d-4f8d-405f-99fb-d62d2c7e01ce-->

andrwng reviewed (commented):
This pretty much LGTM. Just a remaining clarifying question about the completion callback change

dotnwat commented:
Failures unit test https://github.com/redpanda-data/redpanda/issues/16561

dotnwat commented:
> new failures in https://buildkite.com/redpanda/redpanda/builds/46902#018e8115-18bc-4a5e-a666-84e734b6968e:
> 
> ```
> "rptest.tests.transactions_test.TxUpgradeTest.upgrade_does_not_change_tx_coordinator_assignment_test"
> ```

This PR doesn't touch any code outside `v/io`
